cargo-shuttle/src/lib.rs

```diff
@@ -153,44 +179,81 @@ impl Shuttle {
             Command::Project(ProjectCommand::Restart(ProjectStartArgs { idle_minutes })) => {
-                self.project_recreate(&self.client()?, idle_minutes).await
+                self.project_recreate(idle_minutes).await
             }
             Command::Project(ProjectCommand::Status { follow }) => {
-                self.project_status(&self.client()?, follow).await
+                self.project_status(follow).await
             }
             Command::Project(ProjectCommand::List { page, limit }) => {
-                self.projects_list(&self.client()?, page, limit).await
+                self.projects_list(page, limit).await
             }
-            Command::Project(ProjectCommand::Stop) => self.project_delete(&self.client()?).await,
+            Command::Project(ProjectCommand::Stop) => self.project_delete().await,
+        };
+
+        for w in self.version_warnings {
+            println!("{w}");
         }
-        .map(|_| CommandOutcome::Ok)
+
+        res
     }
 
-    fn client(&self) -> Result<Client> {
-        let mut client = Client::new(self.ctx.api_url());
-        client.set_api_key(self.ctx.api_key()?);
-        Ok(client)
+    async fn check_api_versions(&mut self) -> Result<()> {
+        let client = self.client.as_ref().unwrap();
+        debug!("Checking API versions");
+        if let Ok(versions) = client.get_api_versions().await {
+            debug!("Got API versions: {versions:?}");
+            self.version_info = Some(versions);
+
+            // check cargo-shuttle version
+            let my_version = &semver::Version::from_str(VERSION).unwrap();
+            let latest_version = &self.version_info.as_ref().unwrap().cargo_shuttle;
+            if my_version != latest_version {
+                let newer_version_exists = my_version < latest_version;
+                let string = if semvers_are_compatible(my_version, latest_version) {
+                    newer_version_exists.then(|| {
+                        format!("Info: A newer version of cargo-shuttle exists ({latest_version}).")
+                    })
+                    // Having a newer but compatible version does not show warning
+                } else {
+                    newer_version_exists.then(||
+                        formatdoc! {"
+                            Warning:
+                                A newer version of cargo-shuttle exists ({latest_version}).
+                                It is recommended to upgrade.
+                                Refer to the upgrading docs: https://docs.shuttle.rs/configuration/shuttle-versions#upgrading-shuttle-version"
+                        }
+                    ).or_else(||
+                        Some(formatdoc! {"
+                            Warning:
+                                Your version of cargo-shuttle ({my_version}) is newer than what the API expects ({latest_version}).
+                                Things might not work until the API is updated."
```

Comment:
```suggestion
                                Unexpected behavior can occur and this indicates a maintenance operation is underway. 
                                Please contact us on the Shuttle discord server if you want to find more."
```

Reply:
I also think about being clearer here and blocking any command while the API is updated, but the message above can also serve as an intermediate step.

Reply:
I don't think we should use version mismatches as reasons to deny requests yet.